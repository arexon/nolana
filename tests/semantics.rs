macro_rules! test_semantics {
    ($name:ident, $source:literal) => {
        #[test]
        fn $name() {
            let allocator = nolana::allocator::Allocator::default();
            let ret = nolana::parser::Parser::new(&allocator, $source).parse();
            let errors = nolana::semantic::SemanticChecker::default().check(&ret.program);
            insta::with_settings!({ omit_expression => true }, {
                insta::assert_debug_snapshot!(errors);
            });
        }
    };
}

test_semantics!(empty_block_expression, "{}");
test_semantics!(filled_block_expression, "{1;};");

test_semantics!(illegal_string_operation_both, "'foo' + 'bar'");
test_semantics!(illegal_string_operation_left, "'foo' == 1");
test_semantics!(illegal_string_operation_right, "1 + 'bar'");
test_semantics!(unequals_string_operation, "'bar' != 'bar'");
test_semantics!(equals_string_operation, "'bar' == 'bar'");

test_semantics!(assigning_context, "context.foo = 0;");

test_semantics!(break_outside_loop, "break;");
test_semantics!(break_inside_loop, "loop(1, {break;});");

test_semantics!(continue_outside_loop, "continue;");
test_semantics!(continue_inside_loop, "loop(1, {continue;});");
