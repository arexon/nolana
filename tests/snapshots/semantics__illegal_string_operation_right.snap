---
source: tests/semantics.rs
---
[
    Diagnostic {
        inner: DiagnosticInner {
            message: "Strings only support `==` and `!=` operators",
            labels: Some(
                [
                    LabeledSpan {
                        label: None,
                        span: SourceSpan {
                            offset: SourceOffset(
                                0,
                            ),
                            length: 9,
                        },
                        primary: false,
                    },
                ],
            ),
            help: None,
            severity: Error,
        },
    },
]
